src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc

```diff
@@ -631,10 +631,19 @@ void Replayer<I>::handle_copy_image(int r) {
 }
 
 template <typename I>
-void Replayer<I>::handle_copy_image_progress(uint64_t offset, uint64_t total) {
-  dout(10) << "offset=" << offset << ", total=" << total << dendl;
+void Replayer<I>::handle_copy_image_progress(uint64_t object_number,
+                                             uint64_t object_count) {
+  dout(10) << "object_number=" << object_number << ", "
+           << "object_count=" << object_count << dendl;
 
-  // TODO
+  std::unique_lock locker{m_lock};
+  if (m_local_mirror_snap_ns.last_copied_object_number >= object_number) {
+    return;
+  }
+
+  m_local_mirror_snap_ns.last_copied_object_number = object_number;
```

Comment:
@dillaman Wouldn't it be possible that `handle_copy_image_progress` was called for object `N+1` while the object `N` was still not copied? If the sync restarted wouldn't it miss the object `N`?

Reply:
I thought it should only call `update_progress` when all objects `<= N` are safe [1] since that's the assumption that the current image-sync logic expects. 

[1] https://github.com/ceph/ceph/blob/master/src/librbd/deep_copy/ImageCopyRequest.cc#L152

Reply:
Indeed. But then why is the condition `if (m_local_mirror_snap_ns.last_copied_object_number >= object_number)` above?

Reply:
I think I put it there as an assumption w/o using a `ceph_assert` since at the worst-case we would just avoid updating our commit position backwards. I can yank it if it's confusing.

Reply:
Yes, it is a little confusing. I'd prefer `assert` if we don't expect it backwards. Or comment that it should never happen. But I am fine with just yanking it too. Whatever you like.

Reply:
I just yanked it